heapster-saw/README.md

```diff
@@ -1,2 +1,301 @@
+
+Implementation of the Heapster type system of separation types inside SAW,
+including a translation to SAW core.
+
+This remainder of this README contains general information about using
+Heapster about the `examples` directory contained here.
+
+## Building
+
+You will need to follow the instructions in the top-level README to download
+or build a SAW binary, of which Heapster is a part.
+
+If you intend to interact with any of Heapster's Coq output, you will also need
+to follow the instructions in the README in the `saw-core-coq` subdirectory.
+Specifically, after installing the dependencies, you will need to run the
+following (from this directory):
+```bash
+cd ../saw-core-coq/coq
+make
+```
+
+## Using Heapster
+
+This section will walk through the process of using Heapster on some code in a
+C file. This will involve generating LLVM bitcode from your file, writing a SAW
+script to type-check your code with Heapster, and writing a Coq file to prove
+things about the generated functional specification(s).
+
+### Generating LLVM bitcode
+
+The input to Heapster is an LLVM bitcode (`.bc`) file. To generate LLVM
+bitcode from a C file, run the following:
+```bash
+clang -emit-llvm -g -c my_file.c
+```
+Be aware that the resulting bitcode may depend on your `clang` version and your
+operating system. In turn, this means the Heapster commands in your SAW script
+and the proofs in your Coq file may also be dependent on how and where the
+bitcode is generated. For this reason, we provide bitcode files for every
+example in the `examples` directory.
+
+### Type-checking using a SAW script
+
+To use Heapster on your generated bitcode, you can either write a SAW script
+(e.g. `my_file.saw`) or start a SAW interactive session. This document will
+use writing a SAW script as an example, but the commands are the same in
+either case.
+
+To see the documentation for any of the commands mentioned here, type `:help`
+followed by the name of the command in a SAW interactive session.
+
+In order to use Heapster commands, you must begin with:
+```
+enable_experimental;
+```
+You can then load your example bitcode file into Heapster using the following:
+```
+env <- heapster_init_env "my_file" "my_file.bc";
+```
+This file will not go into detail about the process of actually type-checking a
+function with Heapster. Instead, we will briefly discuss a few of the main
+commands used, as well as some examples.
+
+One of the simplest Heapster commands is `heapster_define_perm`, which defines
+a new named permission which can then be used in Heapster types. As an
+example, here's a permission which describes an 64-bit integer value:
+```
+heapster_define_perm
+  env
+  "int64"
+  " "
+  "llvmptr 64"
+  "exists x:bv 64.eq(llvmword(x))";
+```
+It's first argument is the Heapster environment, the second is its name, the
+third is its arguments (of which there are none), the fourth is its type, and
```

Comment:
How about something like "the fourth is the type of value that the permission applies to" or something like that?